### PR TITLE
Ui changes

### PR DESCRIPTION
**Pull request by mrerrormessage**
I'm also planning to try to make the build reproducible and to add the run button today, but I figured I would go ahead and open a PR.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ project
 lib
 target
 *.jar
+*.swp
```
Comment (qiemem) [inline .gitignore:5]: FYI, you can stop vim from polluting your directories in general:

```
set backupdir=~/.vim/backup//
set directory=~/.vim/swap//
set undodir=~/.vim/undo//
```

File: Install.sbt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    val installationFile = installationDir / createdFile.getName
+    if (installationFile.exists) {
+      IO.delete(installationFile)
+    }
+    IO.copy(Seq((createdFile -> installationFile)))
+    println(s"latest LSWidgets installed to ${installationFile.getPath}")
+  } else {
+    println(s"could not find /Applications/NetLogo $version/")
+  }
+}
```
Comment (qiemem) [inline Install.sbt]: I think I would prefer for this not to be in here. It's not cross-platform. Besides, why not just symlink?

File: src/main/scala/LSWidgets.scala [hunk 3]
```diff
@@ -52,73 +54,40 @@ class ProcedureWidget(val key: WidgetKey, val state: State, val ws: GUIWorkspace
...
+  val nameField: JTextField = boundTextField(10, kind.nameProperty)
+  val argField: JTextField = boundTextField(0, kind.argProperty)
+
+  removeAll()
+  setLayout(new BoxLayout(this, BoxLayout.Y_AXIS))
+  add(buttonPanel)
+  add(Box.createRigidArea(new Dimension(0, 3)))
+  add(inputPanel)
+
+  def buttonPanel: JPanel = {
```
Comment (qiemem) [inline src/main/scala/LSWidgets.scala]: Shouldn't this be a `val`, not a `def`? Right now, calling `buttonPanel` multiple times will create a new panel every time.

File: src/main/scala/LSWidgets.scala [hunk 4]
```diff
@@ -52,73 +54,40 @@ class ProcedureWidget(val key: WidgetKey, val state: State, val ws: GUIWorkspace
...
-  add(argField)
-
-  val scrollPane = new JScrollPane(editor)
-  springLayout.putConstraint(NORTH, scrollPane, bigSpace, SOUTH, argLabel)
-  springLayout.putConstraint(NORTH, scrollPane, bigSpace, SOUTH, argField)
-  springLayout.putConstraint(WEST, scrollPane, bigSpace, WEST, this)
-  springLayout.putConstraint(SOUTH, scrollPane, -bigSpace, SOUTH, this)
-  springLayout.putConstraint(EAST, scrollPane, -bigSpace, EAST, this)
-  add(scrollPane)
+  def inputPanel: JPanel = {
```
Comment (qiemem) [inline src/main/scala/LSWidgets.scala]: `val` not `def`.

Comment (qiemem) [inline src/main/scala/LSWidgets.scala]: `val` not `def`

File: src/main/scala/LSWidgets.scala [hunk 5]
```diff
@@ -126,6 +95,13 @@ class ProcedureWidget(val key: WidgetKey, val state: State, val ws: GUIWorkspace
       override def removeUpdate(e: DocumentEvent): Unit = updateInState(property)
       override def insertUpdate(e: DocumentEvent): Unit = updateInState(property)
     })
+
+  private def boundTextField(i: Int, prop: StringProperty[this.type]) = {
+    val t = new JTextField(i)
```
Comment (qiemem) [inline src/main/scala/LSWidgets.scala]: Why set the number of columns at all? I did that for the name field to limit its width, but that's not necessary anymore.

**Comment by qiemem** [comment]
This is great in general. It may be simpler to extract out the argument panels into their own class (`ArgumentPanel`) to better encapsulate their logic.

Also, the argument panels seem to be padded with quite a bit of extra vertical space, even when empty.

File: .gitignore (hunk 1 above)
Comment (mrerrormessage) [inline .gitignore:5]: I'll add that into my vimrc. I still think it's worth having in this file.

File: Install.sbt (hunk 2 above)
Comment (mrerrormessage) [inline Install.sbt]: I'll remove this. ~~A symlink is a much better idea.~~ I was thinking about getting the build reproducible at the time.

EDIT: I have tried symlinking between the extension directory and the project root, and it does not work on my machine. Does it work for you? I can still remove this in git, but I'm planning to keep it locally.

File: src/main/scala/LSWidgets.scala (hunk 3 above)
Comment (mrerrormessage) [inline src/main/scala/LSWidgets.scala]: I'll change def to val everywhere it's appropriate.

File: src/main/scala/LSWidgets.scala (hunk 5 above)
Comment (mrerrormessage) [inline src/main/scala/LSWidgets.scala]: I wasn't sure why you had done it, and I didn't want to introduce a change without a good reason. I'll just remove it.

**Comment by mrerrormessage** [comment]
I'll make the requested changes and see if I can extract the argument panels into their own class. I made a preliminary effort to do so initially. Part of the difficulty is that the xw properties have type parameters which make the type signature of collaborator class methods difficult to get right.

File: Install.sbt (hunk 2 above)
Comment (qiemem) [inline Install.sbt]: It must definitely works for me. You need to symlink from the xw's `widgets` directory to this project's root.

```
▶ pwd
/Users/bryan/Development/NetLogo/eXtraWidgets/xw/widgets

▶ ls -l LSWidgets
lrwxr-xr-x  1 bryan  staff    54B Mar 31 12:46 LSWidgets -> /Users/bryan/Development/NetLogo/xw-widgets/ls-widgets
```

File: .gitignore (hunk 1 above)
Comment (qiemem) [inline .gitignore:5]: Most definitely! Just thought you might find that useful as well.

File: Install.sbt (hunk 2 above)
Comment (mrerrormessage) [inline Install.sbt]: Ah yes, classic `ln -s . /path/to/link` != `ln -s `pwd` /path/to/link`

Comment (qiemem) [inline Install.sbt]: Seriously. Every damn time.

**Comment by mrerrormessage** [comment]
Thanks for the review @qiemem . I believe all changes have been made, please re-review at your convenience. Would you like me to rebase down into fewer commits?

File: src/main/scala/LSWidgets.scala [hunk 6]
```diff
@@ -180,95 +168,110 @@ class Relationship(val key: WidgetKey, val state: State, val ws: GUIWorkspace) e
...
+  def selectedAgentsetArguments_=(args: LogoList): Unit = agentsetArgumentPanel.selectedArguments = args
+
+  def availableAgentsetArguments: LogoList = agentsetArgumentPanel.availableArguments
+  def availableAgentsetArguments_=(args: LogoList) = agentsetArgumentPanel.availableArguments = args
+
+  def selectedProcedureArguments: LogoList = procedureArgumentPanel.selectedArguments
+  def selectedProcedureArguments_=(args: LogoList): Unit = procedureArgumentPanel.selectedArguments = args
+
+  def availableProcedureArguments: LogoList = procedureArgumentPanel.availableArguments
+  def availableProcedureArguments_=(args: LogoList) = procedureArgumentPanel.availableArguments = args
```
Comment (qiemem) [inline src/main/scala/LSWidgets.scala]: You don't need proxy accessors anymore.  Just toss them and directly access the `LSArgumentSelector` accessors in the properties.

Comment (mrerrormessage) [inline src/main/scala/LSWidgets.scala]: Yeah, should have noticed that the binding from xw to the properties was controlled through the widgetKind, not reflection on the widget. Not that I thought it was reflection, I just glossed over the detail.

**Comment by qiemem** [comment]
Looks good overall. The argument panels still seem to be taking up quite a bit of space (especially when empty). Anything you can do about that?

**Comment by mrerrormessage** [comment]
I forgot to adjust that. I'll see if I can get the space down.

**Pull request merged**